include/arch/arm/arch/types.h

```diff
@@ -15,14 +17,21 @@ typedef signed long sword_t;
 /* for printf() formatting */
 #define _seL4_word_fmt  l
 
-typedef word_t vptr_t;
```

Comment:
`vptr_t` is currently used as both user pointers (`ipcbuf_vptr`, `v_entry`, `bi_frame_vptr`, all Arm's cache operations, etc) and offsets/integers as you mentioned. When we need a type to be able to hold both pointers and integers, we just a CHERI type, hence I changed `vptr_t`. 

We could split that up into two types if you'd like, one that's always a pointer, and another that's just an integer (though, I wouldn't include the `ptr_t` part in it because it's not really a pointer). However, that will require changes to existing function definitions, calls, and structures that might be more disruptive than my current changes.

Reply:
Ok your right that sometimes the kernel uses vptr_t as offset into its own address space. I'm not sure whether this goes beyond the cache maintenance instructions though.

Do the cache maintenance operations that operate by VA change on Arm with CHERI to take a capability as the virtual address argument or is it just the address?

There is a place where the kernel does use a vptr_t value to initialize the initial thread's R0 for the address to the bootinfo struct. But it casts the vptr_t to a pointer first. This seems like a valid point to need to explicitly construct a new CHERI capability for launching this thread.

Reply:
> Ok your right that sometimes the kernel uses vptr_t as offset into its own address space. I'm not sure whether this goes beyond the cache maintenance instructions though.
> 
> Do the cache maintenance operations that operate by VA change on Arm with CHERI to take a capability as the virtual address argument or is it just the address?
> 
Yes, they need a valid tagged capability in C64 mode. But in hybrid kernel it's not required.

> There is a place where the kernel does use a vptr_t value to initialize the initial thread's R0 for the address to the bootinfo struct. But it casts the vptr_t to a pointer first. This seems like a valid point to need to explicitly construct a new CHERI capability for launching this thread.

Yeah, bootinfo passed to user, IPC buffer pointer (inside bootinfo struct), and user v_entry (written in the root thread's PC register). So 3 places at least where `vptr_t` is used a pointer and needs to be a valid CHERI capability in a hybrid kernel.

Reply:
But because it's the kernel managing CHERI pointers for a remote address space doesn't it make sense to pass around the raw address space offset and only construct the CHERI pointer right at the end when writing it into the user state?